### PR TITLE
Fix MPS with sequence loss 

### PR DESCRIPTION
**Pull request by JAEarly**
# What does this PR do?

**TLDR**
Fixes the trainer eval loop for MPS devices when the model output is a list of tensors rather than a `torch.Tensor` or `Mapping`.

**Related**
Similar issue but with dict: #2632 
PR for above issue: #2706

**Details**
In the trainer evaluation loop, model outputs (`self.state.outputs`) are moved to the CPU if they are on an MPS device (to avoid torchvision numerical errors on MPS devices). Currently this works fine if `self.state.outputs` is a Mapping or a torch.Tensor. However, it fails for Sequence types (e.g. a list of torch.Tensors).

According to the State class, `outputs` is expected to be of type `torch.Tensor | Sequence[torch.Tensor]`. So Sequence types should be supported in this process in the eval loop. Strictly, `outputs` should not be a `Mapping`, despite the eval operation supporting this. I have left the `Mapping` support in place for now but happy to revisit.

As it was a little difficult to debug this issue, I have added an error message which should make it clearer if an invalid output type is used (such that it cannot correctly be mapped from MPS to CPU). 

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by JAEarly** [comment]
Upon further inspection, `eval_forward` can strictly return `Any` type, so keeping support for `Mapping` makes sense. Adding support for `self.state.options == None` in this PR may also be useful.

**Comment by JAEarly** [comment]
@mvpatel2000 this is a relatively small change but is fairly annoying for running stuff on MPS, could it be merged in please? Thank you!

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -2983,8 +2983,17 @@ def _eval_loop(
...
+                                        outputs = []
+                                        for v in self.state.outputs:
+                                            if isinstance(v, torch.Tensor):
+                                                outputs.append(v.cpu())
+                                            else:
+                                                outputs.append(v)
+                                    elif isinstance(self.state.outputs, torch.Tensor):
                                         outputs = self.state.outputs.cpu()
+                                    else:
+                                        raise ValueError('Unsupported type for state.outputs in trainer eval loop.')
```
Comment (dakinggg) [inline composer/trainer/trainer.py:2996]: To be a bit safer, I think we don't want to error here. Since the output of `eval_forward` is `Any`, the user could actually be returning a custom object (that could in theory have `.cpu()` defined on it. We don't want to break any existing workflows doing something like this. So I would leave the fallback case of `outputs = self.state.outputs.cpu()`. Does this make sense to you?

**Review by dakinggg — COMMENTED** [review]
Thanks for the fix! I left one small comment, and happy to approve if that makes sense to you!

File: composer/trainer/trainer.py (hunk 1 above)
Comment (JAEarly) [reply composer/trainer/trainer.py]: Made the change as you suggested, thanks!

**Review by dakinggg — APPROVED** [review]
Thanks!

**Comment by mvpatel2000** [comment]
@JAEarly apologies for the delay in review! Next time please feel free to tag me as a reviewer.... I will see if we can setup auto-tagging as well. We didn't see this earlier :(

**Review by mvpatel2000 — APPROVED** [review]
